kernel/src/utilities/machine_register.rs

```diff
@@ -0,0 +1,34 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2024.
+
+//! Machine-specific register-sized type.
+//!
+//! This type holds exactly one machine register of data. This type should be
+//! used when storing a value that is exactly the contents of a register with no
+//! additional type information.
+//!
+//! Tock defines this as a custom type as there is currently (Nov 2024) no
+//! suitable standard type for this purpose that is correct across all hardware
+//! architectures Tock supports. The closest suitable type is `usize`. However,
+//! `usize` only captures the size of data, not necessarily the full size of a
+//! register. On many platforms these are the same, but on platforms with
+//! ISA-level memory protection (e.g., CHERI), a register is larger than
+//! `usize`.
+
+use crate::utilities::capability_ptr::CapabilityPtr;
+
+/// [`MachineRegister`] is a datatype that can hold exactly the contents of a
+/// register with no additional semantic information.
+///
+/// We can define this as a [`CapabilityPtr`] as [`CapabilityPtr`] has the same
+/// size.
+///
+/// [`MachineRegister`] is useful for identifying when data within the Tock
+/// kernel has no semantic meaning other than being the size of a register. In
+/// the future it may be possible, useful, or necessary to change the
+/// implementation of [`MachineRegister`], however, the semantics will remain.
+/// To reduce implementation complexity we simply use a type alias, but this
+/// should not be construed as to suggest that [`MachineRegister`] will always
+/// be implemented as a [`CapabilityPtr`].
+pub type MachineRegister = CapabilityPtr;
```

Comment:
_n.b., This comment is non-blocking_

We discussed this as the plan already on the prior capability PR, but looking at it, it's a bit weird to do this. Wouldn't it make more sense to go the other way? i.e., define the `MachineRegister` type precisely on its own as some concrete, sized type. Then, update the `CapabilityPtr` definition to be an alias of the `MachineRegister` type? Semantically, that makes a bit more sense I think — i.e., a `MachineRegister` has a obvious, clear, architectural definition and then a `CapabilityPtr` _adds_ semantics to the `MachineRegister`?

Reply:
Pinging folks who know CHERI better than me (@alevy @LawrenceEsswood @jrvanwhy) — is there a reason not to define this the other way? Not pushing strongly here, just looking to learn.

Reply:
IMO the idiomatic way to do this is with a newtype rather than a type alias. There are two ways you could implement the newtype.

A:
```rust
/// A pointer to userspace memory with implied authority.
pub struct CapabilityPtr { ... }

/// A datatype that can hold exactly the contents of a
/// register with no additional semantic information.
pub struct MachineRegister {
    // We use CapabilityPtr because its implementation
    // happens to be able to hold an arbitrary register value.
    value: CapabilityPtr,
}
```

B:
```rust
/// A datatype that can hold exactly the contents of a
/// register with no additional semantic information.
pub struct MachineRegister { ... }

/// A pointer to userspace memory with implied authority.
pub struct CapabilityPtr {
    value: MachineRegister,
}
```

(I omitted a lot of details of course, like `#[derive(...)]`, `#[repr(...)]`, and a lot of conversions).

Personally, I am a fan of B, because in A `MachineRegister` depends on putting non-pointer values into a `CapabilityPtr`, which ignores the purpose of `CapabilityPtr`.

If we do go with a type alias, I agree with Pat that it would make more sense to go the other way. However, it doesn't matter technically (with a type alias, the types are completely interchangeable) and would require a bigger refactoring to do so.